### PR TITLE
Docs2: Handle new docs entries in the preview

### PR DESCRIPTION
**Pull request by tmeasday**
Telescoping on https://github.com/storybookjs/storybook/pull/18023

- Ensure we set `moduleExport` when processing/preparing a story
- Refactor PreviewWeb to read new entry types from the index to decide on docs only behaviour
- Add `loadDocsEntry` function that pre-loads `storyImports`
- Move `DocsRenderer` to `addon-docs` and rationalize import behaviour
- Add `of={}` behaviour to `Story` element

### CODE REVIEW

File: lib/store/src/StoryStore.ts [hunk 1]
```diff
@@ -61,6 +62,11 @@ export class StoryStore<TFramework extends AnyFramework> {
 
   resolveInitializationPromise: () => void;
 
+  /**
+   * A map of exported ref to story id, for later consumption
+   */
+  refMap: Map<any, StoryId> = new Map();
```
Comment (tmeasday) [inline lib/store/src/StoryStore.ts:68]: `exportMap`

File: lib/preview-web/src/DocsRender.ts [hunk 2]
```diff
@@ -55,21 +76,69 @@ export class DocsRender<TFramework extends AnyFramework> implements Render<TFram
           ...this.store.getStoryContext(renderedStory),
           viewMode: 'docs' as ViewMode,
         } as StoryContextForLoaders<TFramework>),
-      // Put all the storyContext fields onto the docs context for back-compat
-      ...(!global.FEATURES?.breakingChangesV7 && this.store.getStoryContext(this.story)),
     };
+  }
+
+  docsContext(
```
Comment (tmeasday) [inline lib/preview-web/src/DocsRender.ts:82]: Reconcile with legacy context

File: lib/preview-web/src/DocsRender.ts [hunk 3]
```diff
@@ -55,21 +76,69 @@ export class DocsRender<TFramework extends AnyFramework> implements Render<TFram
...
+    const { id, title, name } = this.entry;
+
+    return {
+      id,
+      title,
+      name,
+
+      storyIdByRef: (ref: any) => this.store.storyIdByRef({ ref }),
+
+      componentStories: () => [],
```
Comment (tmeasday) [inline lib/preview-web/src/DocsRender.ts:94]: Make this the set of stories that came from `storyImports`

Comment (tmeasday) [reply lib/preview-web/src/DocsRender.ts:94]: Use a second entry here of `preloadedStories` or similar.

File: lib/preview-web/src/types.ts [hunk 4]
```diff
@@ -27,16 +32,11 @@ export interface DocsContextProps<TFramework extends AnyFramework = AnyFramework
    */
   mdxStoryNameToKey?: Record<string, string>;
   mdxComponentAnnotations?: any;
-
-  // These keys are deprecated and will be removed in v7
```
Comment (shilman) [inline lib/preview-web/src/types.ts:31]: Nice one! :broom:

**Review by shilman — COMMENTED** [review]
@tmeasday Not sure how to best review this one. I agree with the changes in principle per our discussions.

There's still one unit test failure to take care of! 🙏

**Comment by tmeasday** [comment]
I would look closely at the changes to:
 - `StoryStore`
 - `PreviewWeb`
 - `DocsRender` and
 - `useStory`.

**Review by shilman — APPROVED** [review]
100% awesome! So many improvements in here

**Pull request merged**